lib/model/requests_test.go

```diff
@@ -286,17 +284,159 @@ func TestRequestVersioningSymlinkAttack(t *testing.T) {
+	ign := "ignoredNonExisting"
+	ignExisting := "ignoredExisting"
+
+	fc.addFile(invIgn, 0644, protocol.FileInfoTypeFile, contents)
+	fc.addFile(invDel, 0644, protocol.FileInfoTypeFile, contents)
+	fc.deleteFile(invDel)
+	fc.addFile(ign, 0644, protocol.FileInfoTypeFile, contents)
+	fc.addFile(ignExisting, 0644, protocol.FileInfoTypeFile, contents)
+	if err := ioutil.WriteFile(filepath.Join(tmpDir, ignExisting), otherContents, 0644); err != nil {
+		panic(err)
+	}
+
+	done := make(chan struct{})
+	fc.mut.Lock()
+	fc.indexFn = func(folder string, fs []protocol.FileInfo) {
+		expected := map[string]struct{}{invIgn: {}, ign: {}, ignExisting: {}}
+		for _, f := range fs {
+			if _, ok := expected[f.Name]; !ok {
+				t.Fatalf("Unexpected file %v was added to index", f.Name)
+			}
+			if !f.Invalid {
+				t.Errorf("File %v wasn't marked as invalid", f.Name)
+			}
+			delete(expected, f.Name)
+		}
+		for name := range expected {
+			t.Errorf("File %v wasn't added to index", name)
+		}
+		done <- struct{}{}
+	}
+	fc.mut.Unlock()
+
+	sub := events.Default.Subscribe(events.FolderErrors)
+	defer events.Default.Unsubscribe(sub)
+
+	fc.sendIndexUpdate()
+
+	timeout := time.NewTimer(5 * time.Second)
+	select {
+	case ev := <-sub.C():
+		t.Fatalf("Errors while pulling: %v", ev)
+	case <-timeout.C:
+		t.Fatalf("timed out before index was received")
+	case <-done:
+		return
+	}
+
+	fc.mut.Lock()
+	fc.indexFn = func(folder string, fs []protocol.FileInfo) {
+		expected := map[string]struct{}{ign: {}, ignExisting: {}}
+		for _, f := range fs {
+			if _, ok := expected[f.Name]; !ok {
+				t.Fatalf("Unexpected file %v was updated in index", f.Name)
+			}
+			if f.Invalid {
+				t.Errorf("File %v is still marked as invalid", f.Name)
+			}
+			// The unignored files should only have a local version,
+			// to mark them as in conflict with any other existing versions.
+			ev := protocol.Vector{}.Update(device1.Short())
+			if v := f.Version; !v.Equal(ev) {
+				t.Errorf("File %v has version %v, expected %v", f.Name, v, ev)
+			}
+			if f.Name == ign {
+				if !f.Deleted {
+					t.Errorf("File %v was not marked as deleted", f.Name)
+				}
+			} else if f.Deleted {
+				t.Errorf("File %v is marked as deleted", f.Name)
+			}
+			delete(expected, f.Name)
+		}
+		for name := range expected {
+			t.Errorf("File %v wasn't updated in index", name)
+		}
+		done <- struct{}{}
+	}
+	// Make sure pulling doesn't interfere, as index updates are racy and
+	// thus we cannot distinguish between scan and pull results.
+	fc.requestFn = func(folder, name string, offset int64, size int, hash []byte, fromTemporary bool) ([]byte, error) {
+		return nil, nil
+	}
+	fc.mut.Unlock()
+
+	if err := m.SetIgnores("default", []string{"*:ignored*"}); err != nil {
+		panic(err)
+	}
+
+	timeout = time.NewTimer(5 * time.Second)
+	select {
+	case <-timeout.C:
+		t.Fatalf("timed out before index was received")
+	case <-done:
+		return
 	}
+}
+
+func TestPullInvalidIgnoredSO(t *testing.T) {
```

Comment:
Nano-nit: put these above the helper function, so the helper function makes sense when encountered. (I had to cmd-f to see if it was actually called).